ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/transaction/TransactionSignatureService.kt

```diff
@@ -0,0 +1,57 @@
+package net.corda.v5.ledger.common.transaction
+
+import net.corda.v5.application.crypto.DigitalSignatureAndMetadata
+import net.corda.v5.application.crypto.DigitalSignatureVerificationService
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.crypto.SecureHash
+import net.corda.v5.crypto.merkle.MerkleProof
+import java.security.PublicKey
+
+/**
+ * TransactionSignatureService can be used to sign and verify transaction signatures.
+ * It supports both single and batch signatures.
+ */
+
+interface TransactionSignatureService {
+    /**
+     * Signs a transaction id with all the available keys.
+     *
+     * @param transactionId The transaction id to be signed.
+     * @param publicKeys The keys which should be tried to sign with.
+     *
+     * @returns Resulting signatures.
+     *
+     * @throws TransactionNoAvailableKeysException If none of the keys are available.
+     */
+    @Suspendable
+    fun sign(transactionId: SecureHash, publicKeys: Iterable<PublicKey>): List<DigitalSignatureAndMetadata>
+
+    /**
+     * Signs a list of transactions with each the available keys.
+     * It creates one batch signature for each available keys.
+     * Then returns the signatures for each transaction with a [MerkleProof] to prove that they are included in the batch.
+     *
+     * @param transactions The transactions to be signed.
+     * @param publicKeys The keys which should be tried to sign with.
+     *
+     * @returns Resulting signatures.
```

Comment:
```suggestion
     * @returns List of signatures for each supplied transaction. The list will always be of the same size and in the same order as the supplied [transactions]
```

People may assume this, but I think it's important to make this guarantee on the API. We should also ensure (in runtime-os) that our unit tests verify this. I've been bitten before with a similar list return interface in the C4 uniqueness provider where we inadvertently broke the ordering in a change.

Reply:
Thanks. I applied it and added an `outer` before the list at the beginning of the second phrase.